### DISCUSSION TITLE
Chunk management system proposal

### DISCUSSION THREAD

**Discussion by AleksyBalazinski** [Non-Euclidean-World/Hyper#21] [Ideas]
# Chunk management system proposal
## Assumptions
* Chunks are stored in two locations: in program memory (cache) and on disk (main storage).
* Cache has a fixed capacity.
    Reason: Keeping the memory usage more or less constant.
* Chunks are saved to main storage only if their corressponding values of scalar field have been modified; otherwise they are generated on the fly.
    Reason: We can save only the modified chunks because so far we've seen that generation is quite fast.
* Chunks are identified only by the position of a specified vertex (for example the one with min. values of x, y & z coordinates) -- this is the chunk's ID.
* Along with chunk ID, the scalar field corresponding to that chunk is stored, and can be retreived given the chunk's ID.
    Reason: We keep the scalar field specificaly because this is needed to apply terrain modifications.

## Rendering
On every render frame:
```
Determine positions of visible chunks v_chunks_pos
for each ch_pos in v_chunks_pos:
    if chunk ch with ID ch_pos is in cache:
        ch.render()
    else if chunk ch with ch_pos is on disk:
        load ch from disk
        cache.add(ch)
        ch.redner()
    else:
        create chunk ch (calculate scalar field)
        cache.add(ch)
        ch.render()
```
## Cache
Cache is a FIFO queue.
Adding chunk ch to cache:
```
if cache is full:
    if cache.peek().isModified:
        dump cache[-1] to disk
    cache.dequeue()
cache.enqueue(ch)
```
## Main storage
To be specified. Document-oriented DB can be considered.

**Comment by KarolDenst** [comment]
I don't know if fifo is an efficient way of doing this. Maybe before rendering just remove the chanks that should not be in cache based on position? I think with fifo there is a chance of removing a chunk and the adding the same chank in the same frame.

**Reply by AleksyBalazinski** [reply]
The unwritten assumption was that the cache would store a few times more chunks than are on average visible and that should at least mitigate the problem if not solve it. I thought about FIFO because it should be faster since it doesn't require looping over stored chunks and determining which ones should be removed (based on distance).

**Reply by KarolDenst** [reply]
I think the time it takes to loop over chunks is negligibly small. Loading and saving chunks probably takes much much more time in comparison
